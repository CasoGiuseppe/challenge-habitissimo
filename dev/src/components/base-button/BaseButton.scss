.base-button {
  $basebutton: &;

  @include text--xsmall;
  @include text--regular;

  background: $c-primary;
  border: none;
  border-radius: px2rem($base * 3) 0px px2rem($base * 3) px2rem($base * 3);
  cursor: pointer;
  color: $c-white;
  min-width: #{$base * 20px};
  min-height: #{$base * 6px};
  padding: 0 #{$base * 2px};
  transition: background 1s $a-ease-out,
              color .6s $a-ease-out;
  text-transform: uppercase;

  &:disabled {
    background: $c-dark-gray;
  }

  @include breakpoint('tablet-pt') {
    &:not([disabled]) {
      &:hover,
      &:focus {
        background: $c-secondary;
      }
    }
  }

  &--isSecondary {
    background: $c-secondary;

    @include breakpoint('tablet-pt') {
      &:not([disabled]) {
        &:hover,
        &:focus {
          background: $c-primary;
        }
      }
    }
  }

  &--isLight {
    background: $c-white;
    border: 1px solid $c-silver;
    color: $c-dark-gray;

    @include breakpoint('tablet-pt') {
      &:not([disabled]) {
        &:hover,
        &:focus {
          background: $c-dark-gray;
          color: $c-white;
        }
      }
    }
  }

  &--isFullsize {
    min-width: 100%;
  }

  &--isReplaced {
    min-width: inherit;

    @include breakpoint('only-phone') {
      min-width: #{$base * 6px};
    }
  }

  &__label {
    #{$basebutton}--isReplaced & {
      @include breakpoint('only-phone') {
        display: none;
      }
    }
  }

  &__icon {
    display: none;
    @include breakpoint('only-phone') {
      display: block;
    }
  }

  &[style*="min-width"] {
    @include breakpoint('only-phone') {
      min-width: 100% !important;
    }
  }
}