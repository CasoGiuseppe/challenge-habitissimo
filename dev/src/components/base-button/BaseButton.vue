<template>
  <button
    :class="[
      'base-button',
      isSecondary ? 'base-button--isSecondary' : null,
      isFullsize ? 'base-button--isFullsize' : null,
    ]"
    :disabled="isDisabled"
    @click="handleClickState">
    <slot />
  </button>
</template>

<script>
export default {
  name: 'BaseButton',

  props: {
    isDisabled: {
      type: Boolean,
      default: false,
    },

    isSecondary: {
      type: Boolean,
      default: false,
    },

    isFullsize: {
      type: Boolean,
      default: false,
    },
  },

  methods: {
    handleClickState(e) {
      this.$emit('click', e);
    },
  },
};
</script>

<style lang='scss' scoped src= './BaseButton.scss'></style>
