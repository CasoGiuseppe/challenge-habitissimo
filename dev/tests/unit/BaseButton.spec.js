import BaseButton from '@/components/base-button/BaseButton'
import { mount } from '@vue/test-utils'

describe('BaseButton.vue', () => {
  let component

  beforeEach(() => {
    component = mount(BaseButton, {
      propsData:{
        isDisabled: true,
        isSecondary: true,
        isFullsize: true,
        isLight: true,
        hasMinWidth: "100px",
        isReplaced: {
          state: false,
          icon: 'svg',
          color: 'primary'
        }
      },
      slots: {
        default: '<span class="slot-action">button</span>'
      }
    })
  })

  it('has correct props values', () => {
    expect(component.props().isDisabled).toBe(true)
    expect(component.props().isSecondary).toBe(true)
    expect(component.props().isFullsize).toBe(true)
    expect(component.props().isLight).toBe(true)
    expect(component.props().hasMinWidth).toBe('100px')
    expect(component.props().isReplaced).toHaveProperty('icon', 'svg')
  })

  it('has correct $emit response', async () => {
    component.vm.$emit('click', 'clickvalue')

    await component.vm.$nextTick()
    expect(component.emitted().click).toBeTruthy()
    expect(component.emitted().click[0]).toEqual(['clickvalue'])
  })

  it('has correct action slot', () => {
    const main = component.find('.base-button')
    expect(main.findAll(".slot-action").length).toBe(1);
  })

  it('has correct class name', () => {
    expect(component.classes()).toContain('base-button--isLight')
  })

  it('has the expected html structure', () => {
    expect(component.html()).toMatchSnapshot()
  })
})