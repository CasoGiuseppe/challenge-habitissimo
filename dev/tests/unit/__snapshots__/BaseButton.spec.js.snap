// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton.vue has the expected html structure 1`] = `
<button disabled="disabled" class="base-button base-button--isSecondary base-button--isLight base-button--isFullsize" style="min-width: 100px;"><span class="base-button__label"><span class="slot-action">button</span></span>
  <!----></button>
`;
